### PR TITLE
Feature/3113181 drupal composer update

### PR DESCRIPTION
**Pull request by Kingdutch**
## Change
Open Social is using the old composer scaffold plugin that will no longer be maintained for Drupal 9. The recommended method is to use the `core-composer-scaffold` package instead. The added benefit of that package is that it will copy files from installed composer packages instead of downloading each file one by one.

The cweagan/installers package is updated for compatibility with composer 2 and new Drupal related installer types that are supported.

## Issue tracker
https://www.drupal.org/project/social/issues/3113181

## How to test
Copy the following composer.json and run `composer install`. Besides this tests should pass. The composer.json below is a slightly modified version of the improved social_template.

```
{
    "name": "goalgorilla/social_template",
    "description": "Social project template for composer based Open Social projects.",
    "type": "project",
    "license": "GPL-2.0+",
    "minimum-stability": "dev",
    "prefer-stable": true,
    "require": {
        "php": "^7.2",
        "goalgorilla/open_social": "dev-feature/3113181-drupal-composer-update",
        "monolog/monolog": "^1.17"
    },
    "repositories": [
        {
            "type": "composer",
            "url": "https://packages.drupal.org/8"
        },
        {
            "type": "composer",
            "url": "https://asset-packagist.org"
        }
    ],
    "extra": {
        "drupal-scaffold": {
            "locations": {
                "web-root": "html/"
            }
        },
        "installer-types": [
            "bower-asset",
            "npm-asset"
        ],
        "installer-paths": {
            "html/core": [
                "drupal/core"
            ],
            "html/modules/contrib/{$name}": [
                "type:drupal-module"
            ],
            "html/profiles/contrib/social": [
                "goalgorilla/open_social"
            ],
            "html/profiles/contrib/{$name}": [
                "type:drupal-profile"
            ],
            "html/themes/contrib/{$name}": [
                "type:drupal-theme"
            ],
            "html/libraries/{$name}": [
                "type:drupal-library",
                "type:bower-asset",
                "type:npm-asset"
            ],
            "drush/contrib/{$name}": [
                "type:drupal-drush"
            ]
        },
        "enable-patching": true
    }
}
```

## Release notes
Open Social now uses the recommended `drupal/core-composer-scaffold` package to download scaffolding files. This replaces the outdated `drupal-composer/drupal-scaffold` package. Please check the change-record for details. 

## Change Record
Please review the following change record.

[ ] Agreed with the remarks about `drupal/drupal` vs `drupal/core` for framing of Open Social's role as package in a composer project.

https://www.drupal.org/node/3154837

### CODE REVIEW

**Comment by Kingdutch** [comment]
The following changes are made to Social Template with this PR in mind. Mostly focusing on deduplicating dependencies: https://github.com/goalgorilla/social_template/pull/24/

File: composer.json [hunk 1]
```diff
@@ -110,10 +109,10 @@
     },
     "require": {
         "cweagans/composer-patches": "^1.6.0",
-        "composer/installers": "^1.6",
+        "composer/installers": "^1.9",
         "oomphinc/composer-installers-extender": "^1.1",
-        "drupal-composer/drupal-scaffold": "^2.5.0",
```
Comment (vredeling) [inline composer.json:115]: When using goalgorilla/social_template there is still a reference to ` "drupal-composer/drupal-scaffold": "^2.6.1"` in there which conflicts with this change. Without the aforementioned conflict composer will install / update nicely with the new scaffold.

Comment (Kingdutch) [reply composer.json:115]: Yep, as mentioned in https://github.com/goalgorilla/open_social/pull/1889#issuecomment-649555596

The PR in https://github.com/goalgorilla/social_template/pull/24 will need to be merged with this PR to make sure that conflict doesn't happen.

**Review by navneet0693 — APPROVED** [review]

**Comment by navneet0693** [comment]
Even though tests seems to be failing here and there, they are mostly unrelated to changes here.

**Comment by navneet0693** [comment]
🍒 picked to 8.x and 9.x

**Comment by Kingdutch** [comment]
We've reverted the cherry-picks to 8.x and 9.x since these composer changes are a backwards compatibility break and require manual changes to a composer.json file to allow updating. This means we can only include this in 10.0.0 (the next major release) :)

**Pull request merged**